File: 05-java-mqtt/src/main/java/org/protelis/demo/HelloProtelis.java
package org.protelis.demo;

import com.uchuhimo.konf.BaseConfig;
import com.uchuhimo.konf.Config;
import com.uchuhimo.konf.source.DefaultTomlLoaderKt;
import org.protelis.demo.data.MqttProtelisNode;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.protelis.lang.ProtelisLoader;
import org.protelis.vm.ProtelisProgram;

import java.util.ArrayList;
import java.util.List;

/**
 * Entrypoint for the MQTT demo.
 */
public final class HelloProtelis {

    private HelloProtelis() { }

    /**
     * Main method.
     * @param args unused.
     */
    public static void main(final String[] args) {
        Config config = new BaseConfig();
        config.addSpec(ProtelisConfigSpec.SPEC);
        config = DefaultTomlLoaderKt.getToml(config.from()).resource("config.toml", false);
        final String protelisModuleName = config.get(ProtelisConfigSpec.protelisModuleName);
        final int iterations = config.get(ProtelisConfigSpec.iterations);
        final List<MqttProtelisNode> nodes = config.get(ProtelisConfigSpec.nodes);
        final List<Device> devices = new ArrayList<>();
        // Initialize each node
        nodes.forEach(n -> {
            final MqttNetworkManager netmgr = new MqttNetworkManager(new IntDeviceUID(n.getId()), n.getNeighbors());
            try {
                netmgr.listen(n.getListen()).waitForCompletion();
            } catch (MqttException e) {
                throw new IllegalStateException(e);
            }
            final ProtelisProgram program = ProtelisLoader.parse(protelisModuleName);
            final Device node = new Device(program, n.getId(), netmgr, new ConsoleSpeaker());
            if (n.isLeader()) {
                node.getDeviceCapabilities().getExecutionEnvironment().put("leader", true);
            }
            devices.add(node);
        });
        // Run some cycles
        for (int i = 0; i < iterations; i++) {
            devices.forEach(Device::runCycle);
        }
        // Stop the server socket
        devices.forEach(d -> {
            try {
                ((MqttNetworkManager) d.getNetworkManager()).stop().waitForCompletion();
            } catch (MqttException e) {
                throw new IllegalStateException(e);
            }
        });
    }
}

File: 06-kotlin-mqtt/src/main/kotlin/HelloProtelis.kt
package demo

import com.uchuhimo.konf.Config
import com.uchuhimo.konf.source.toml
import org.protelis.lang.ProtelisLoader

/**
 * Demonstration of the [MqttNetworkManager] usage.
 */
fun main() {
    var devices: List<Device> = emptyList()
    val config = Config { addSpec(ProtelisConfigSpec) }
            .from.toml.resource("config.toml")
    val protelisModuleName = config[ProtelisConfigSpec.protelisModuleName]
    val iterations = config[ProtelisConfigSpec.iterations]
    val nodes = config[ProtelisConfigSpec.nodes]
    // Initialize some nodes.
    nodes.forEach {
        val mqttNetworkManager = MqttNetworkManager(IntDeviceUID(it.id), it.neighbors).apply { listen(it.listen) }
        val program = ProtelisLoader.parse(protelisModuleName)
        val node = Device(program, it.id, mqttNetworkManager, ConsoleSpeaker())
        if (it.leader) {
            node.deviceCapabilities.executionEnvironment.put("leader", true)
        }
        devices += node
    }
    // Let the nodes make some iterations.
    repeat(iterations) {
        devices.forEach { it.runCycle() }
    }
    // Close the thread listening.
    devices.forEach { (it.netmgr as MqttNetworkManager).stop() }
}